File: experiments/engine-kt/provider-compatible/src/main/kotlin/com/garive/eng/kt/provider/compatible/OutcomeMapping.kt
package com.garive.eng.kt.provider.compatible

import com.garive.eng.kt.anthropic.MessageResponse
import com.garive.eng.kt.anthropic.OutputBlock as MessagesOutput
import com.garive.eng.kt.anthropic.StopReason
import com.garive.eng.kt.anthropic.Usage as MessagesUsage
import com.garive.eng.kt.llm.InterruptionKind
import com.garive.eng.kt.llm.InvokeOutcome
import com.garive.eng.kt.llm.ModelItem
import com.garive.eng.kt.llm.ModelStopReason
import com.garive.eng.kt.llm.ModelUsage
import com.garive.eng.kt.llm.ReasoningContent
import com.garive.eng.kt.llm.RejectionKind
import com.garive.eng.kt.llm.TokenCount
import com.garive.eng.kt.llm.UsageSource
import com.garive.eng.kt.openai.OutputContent as ResponsesContent
import com.garive.eng.kt.openai.ResponseEnvelope
import com.garive.eng.kt.openai.ResponseOutputItem
import com.garive.eng.kt.openai.ResponseStatus
import com.garive.eng.kt.openai.ResponseUsage
import kotlin.time.Duration
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject

/** Classifies an exact typed protocol error without inspecting its message. */
public fun classifyProtocolError(
    policy: ProtocolErrorPolicy,
    signature: ErrorSignature,
    retryAfter: Duration?,
): InvokeOutcome = when (val disposition = policy.classify(signature)
    ?: fail(CompatibleProviderError.UNCLASSIFIED_PROTOCOL_ERROR)) {
    is ErrorDisposition.Rejected -> InvokeOutcome.Rejected(
        disposition.kind,
        "status=${signature.status};type=${signature.protocolType};code=${signature.code.orEmpty()}",
    )
    is ErrorDisposition.Unavailable -> InvokeOutcome.Unavailable(disposition.kind, retryAfter)
    is ErrorDisposition.Interrupted -> InvokeOutcome.Interrupted(disposition.kind, emptyList(), unknownUsage())
}

/** Normalizes one adapter-validated buffered Responses terminal. */
public fun normalizeResponses(response: ResponseEnvelope, reasoningVisibility: Boolean): InvokeOutcome {
    val items = responsesItems(response.output, reasoningVisibility)
    val usage = response.usage?.let(::responsesUsage) ?: unknownUsage()
    return when {
        response.status == ResponseStatus.COMPLETED && response.error == null -> InvokeOutcome.Completed(
            items,
            usage,
            when {
                items.any { it is ModelItem.ToolIntent } -> ModelStopReason.ToolUse
                items.any { it is ModelItem.Refusal } -> ModelStopReason.Refusal
                else -> ModelStopReason.EndTurn
            },
        )
        response.status == ResponseStatus.INCOMPLETE && response.error == null &&
            response.incompleteDetails?.reason == "max_output_tokens" -> InvokeOutcome.Interrupted(
            InterruptionKind.OUTPUT_LIMIT,
            items,
            usage,
        )
        response.status == ResponseStatus.CANCELLED && response.error == null -> InvokeOutcome.Interrupted(
            InterruptionKind.CANCELLED,
            items,
            usage,
        )
        else -> fail(CompatibleProviderError.PROTOCOL_INVARIANT)
    }
}

/** Normalizes one adapter-validated buffered Messages terminal. */
public fun normalizeMessages(response: MessageResponse, reasoningVisibility: Boolean): InvokeOutcome {
    val refusal = response.stopReason == StopReason.REFUSAL
    val items = response.content.map { messagesItem(it, reasoningVisibility, refusal) }
    return when (val reason = response.stopReason ?: fail(CompatibleProviderError.PROTOCOL_INVARIANT)) {
        StopReason.MAX_TOKENS -> InvokeOutcome.Interrupted(InterruptionKind.OUTPUT_LIMIT, items, messagesUsage(response.usage))
        StopReason.MODEL_CONTEXT_WINDOW_EXCEEDED -> InvokeOutcome.Rejected(
            RejectionKind.CONTEXT_OVERFLOW,
            "model_context_window_exceeded",
        )
        else -> InvokeOutcome.Completed(items, messagesUsage(response.usage), messagesStop(reason))
    }
}

internal fun responsesItems(
    output: List<ResponseOutputItem>,
    reasoningVisibility: Boolean,
): List<ModelItem> = buildList {
    output.forEach { item ->
        when (item) {
            is ResponseOutputItem.Message -> item.content.forEach { content ->
                add(when (content) {
                    is ResponsesContent.Text -> ModelItem.Text(content.text)
                    is ResponsesContent.Refusal -> ModelItem.Refusal(content.refusal)
                    is ResponsesContent.Extension -> fail(CompatibleProviderError.UNSUPPORTED_EXTENSION)
                })
            }
            is ResponseOutputItem.FunctionCall -> add(ModelItem.ToolIntent(
                item.callId,
                item.name,
                canonicalObject(item.arguments),
            ))
            is ResponseOutputItem.Reasoning -> {
                if (reasoningVisibility) {
                    val parts = item.content ?: item.summary
                    if (parts.isNotEmpty()) add(ModelItem.Reasoning(ReasoningContent.ModelVisible(parts.joinToString("") { it.text })))
                } else {
                    item.encryptedContent?.let { add(ModelItem.Reasoning(ReasoningContent.OpaqueReference(it))) }
                }
            }
            is ResponseOutputItem.Extension -> fail(CompatibleProviderError.UNSUPPORTED_EXTENSION)
        }
    }
}

internal fun messagesItem(block: MessagesOutput, reasoningVisibility: Boolean, refusal: Boolean): ModelItem =
    when (block) {
        is MessagesOutput.Text -> if (refusal) ModelItem.Refusal(block.text) else ModelItem.Text(block.text)
        is MessagesOutput.Thinking -> ModelItem.Reasoning(
            if (reasoningVisibility) ReasoningContent.ModelVisible(block.thinking)
            else ReasoningContent.OpaqueReference(block.signature),
        )
        is MessagesOutput.RedactedThinking -> ModelItem.Reasoning(ReasoningContent.OpaqueReference(block.data))
        is MessagesOutput.ToolUse -> ModelItem.ToolIntent(block.id, block.name, block.input.toString())
        is MessagesOutput.Extension -> fail(CompatibleProviderError.UNSUPPORTED_EXTENSION)
    }

internal fun messagesStop(reason: StopReason): ModelStopReason = when (reason) {
    StopReason.END_TURN -> ModelStopReason.EndTurn
    StopReason.STOP_SEQUENCE -> ModelStopReason.StopSequence
    StopReason.TOOL_USE -> ModelStopReason.ToolUse
    StopReason.PAUSE_TURN -> ModelStopReason.PauseTurn
    StopReason.REFUSAL -> ModelStopReason.Refusal
    StopReason.MAX_TOKENS, StopReason.MODEL_CONTEXT_WINDOW_EXCEEDED -> fail(CompatibleProviderError.PROTOCOL_INVARIANT)
}

private fun canonicalObject(encoded: String): String = try {
    (Json.parseToJsonElement(encoded) as? JsonObject)?.toString()
        ?: fail(CompatibleProviderError.PROTOCOL_INVARIANT)
} catch (error: CompatibleProviderException) {
    throw error
} catch (_: IllegalArgumentException) {
    fail(CompatibleProviderError.PROTOCOL_INVARIANT)
}

private fun responsesUsage(value: ResponseUsage): ModelUsage = ModelUsage(
    TokenCount.Known(value.inputTokens),
    TokenCount.Known(value.outputTokens),
    TokenCount.Known(value.inputTokensDetails.cachedTokens),
    TokenCount.Known(value.inputTokensDetails.cacheWriteTokens),
    UsageSource.PROVIDER_REPORTED,
)

internal fun messagesUsage(value: MessagesUsage): ModelUsage = ModelUsage(
    TokenCount.Known(value.inputTokens),
    TokenCount.Known(value.outputTokens),
    value.cacheReadInputTokens?.let(TokenCount::Known),
    value.cacheCreationInputTokens?.let(TokenCount::Known),
    UsageSource.PROVIDER_REPORTED,
)

internal fun unknownUsage(): ModelUsage = ModelUsage(
    TokenCount.Unknown,
    TokenCount.Unknown,
    source = UsageSource.PROVIDER_REPORTED,
)

internal fun fail(error: CompatibleProviderError): Nothing = throw CompatibleProviderException(error)

File: providers/compatible/tests/outcome_mapping.rs
use std::time::Duration;

use garive_llm::{
    InterruptionKind, InvokeOutcome, ModelItem, ModelStopReason, RejectionKind, TokenCount,
    UnavailableKind,
};
use garive_provider_compatible::{
    classify_protocol_error, normalize_messages, normalize_responses, CompatibleProviderError,
    ErrorDisposition, ErrorSignature, ProtocolErrorPolicy,
};
use serde_json::{json, Value};

fn fixture() -> Value {
    serde_json::from_str(include_str!(
        "../../../spec/fixtures/providers/compatible-mapping-v1.json"
    ))
    .expect("valid shared fixture")
}

fn responses_wire(case: &Value) -> Value {
    let source = &case["response"];
    let output = source["items"]
        .as_array()
        .expect("items")
        .iter()
        .enumerate()
        .map(|(index, item)| match item["kind"].as_str().expect("item kind") {
            "text" => json!({
                "type": "message", "id": format!("msg-{index}"), "role": "assistant",
                "status": "completed", "content": [{"type": "output_text", "text": item["text"], "annotations": []}]
            }),
            "tool" => json!({
                "type": "function_call", "call_id": item["model_call_id"],
                "name": item["tool_name"], "arguments": item["arguments"].to_string(), "status": "completed"
            }),
            other => panic!("unsupported fixture item {other}"),
        })
        .collect::<Vec<_>>();
    let usage = source["usage"].as_object().map(|usage| {
        let input = usage["input"].as_u64().expect("input usage");
        let output = usage["output"].as_u64().expect("output usage");
        json!({
            "input_tokens": input,
            "input_tokens_details": {"cached_tokens": usage["cache_read"], "cache_write_tokens": usage["cache_write"]},
            "output_tokens": output,
            "output_tokens_details": {"reasoning_tokens": 0},
            "total_tokens": input + output
        })
    });
    json!({
        "id": "response-1", "created_at": 1.0, "error": null,
        "incomplete_details": source.get("reason").filter(|value| !value.is_null()).map(|reason| json!({"reason": reason})),
        "instructions": null, "metadata": null, "model": "fixture", "object": "response",
        "output": output, "parallel_tool_calls": false, "temperature": null, "text": null,
        "tool_choice": "auto", "tools": [], "top_p": null, "status": source["status"], "usage": usage
    })
}

fn messages_wire(case: &Value) -> Value {
    let source = &case["response"];
    let content = source["items"]
        .as_array()
        .expect("items")
        .iter()
        .map(|item| match item["kind"].as_str().expect("item kind") {
            "refusal" | "text" => json!({"type": "text", "text": item["text"]}),
            other => panic!("unsupported fixture item {other}"),
        })
        .collect::<Vec<_>>();
    let usage = source["usage"].as_object().expect("messages usage");
    json!({
        "id": "message-1", "type": "message", "role": "assistant", "model": "fixture",
        "content": content, "stop_reason": source["stop_reason"], "stop_sequence": null,
        "usage": {"input_tokens": usage["input"], "output_tokens": usage["output"],
            "cache_read_input_tokens": usage["cache_read"], "cache_creation_input_tokens": usage["cache_write"]}
    })
}

#[test]
fn shared_responses_terminals_preserve_items_usage_and_interruption() {
    let fixture = fixture();
    let completed: garive_openai_responses::Response =
        serde_json::from_value(responses_wire(&fixture["outcome_cases"][0])).expect("response");
    let outcome = normalize_responses(&completed, false).expect("normalize completed");
    let InvokeOutcome::Completed {
        items,
        usage,
        stop_reason,
    } = outcome
    else {
        panic!("completed outcome")
    };
    assert_eq!(stop_reason, ModelStopReason::ToolUse);
    assert!(matches!(items[0], ModelItem::Text { .. }));
    assert!(matches!(items[1], ModelItem::ToolIntent { .. }));
    assert_eq!(usage.input_tokens, TokenCount::Known(10));
    assert_eq!(usage.output_tokens, TokenCount::Known(4));

    let incomplete: garive_openai_responses::Response =
        serde_json::from_value(responses_wire(&fixture["outcome_cases"][1])).expect("response");
    assert!(matches!(
        normalize_responses(&incomplete, false).expect("normalize incomplete"),
        InvokeOutcome::Interrupted {
            kind: InterruptionKind::OutputLimit,
            ..
        }
    ));
}

#[test]
fn shared_messages_terminals_distinguish_refusal_and_context_rejection() {
    let fixture = fixture();
    let refusal: garive_anthropic_messages::MessageResponse =
        serde_json::from_value(messages_wire(&fixture["outcome_cases"][2])).expect("message");
    let outcome = normalize_messages(&refusal, false).expect("normalize refusal");
    assert!(matches!(
        outcome,
        InvokeOutcome::Completed {
            stop_reason: ModelStopReason::Refusal,
            ref items,
            ..
        } if matches!(items[0], ModelItem::Refusal { .. })
    ));

    let context: garive_anthropic_messages::MessageResponse =
        serde_json::from_value(messages_wire(&fixture["outcome_cases"][3])).expect("message");
    assert!(matches!(
        normalize_messages(&context, false).expect("normalize context"),
        InvokeOutcome::Rejected {
            kind: RejectionKind::ContextOverflow,
            ..
        }
    ));
}

#[test]
fn shared_error_cases_use_only_exact_signatures() {
    let fixture = fixture();
    let policy = ProtocolErrorPolicy::new([
        (
            ErrorSignature {
                status: 401,
                protocol_type: "authentication_error".into(),
                code: None,
            },
            ErrorDisposition::Rejected(RejectionKind::Authentication),
        ),
        (
            ErrorSignature {
                status: 429,
                protocol_type: "rate_limit_error".into(),
                code: Some("rate_limit".into()),
            },
            ErrorDisposition::Unavailable(UnavailableKind::RateLimited),
        ),
    ])
    .expect("unique policy");

    for case in fixture["error_cases"].as_array().expect("error cases") {
        let signature = ErrorSignature {
            status: case["status"].as_u64().expect("status") as u16,
            protocol_type: case["type"].as_str().expect("type").to_owned(),
            code: case["code"].as_str().map(str::to_owned),
        };
        let result = classify_protocol_error(&policy, signature, Some(Duration::from_secs(2)));
        match case["expected"].as_str().expect("expected") {
            "authentication" => assert!(matches!(
                result,
                Ok(InvokeOutcome::Rejected {
                    kind: RejectionKind::Authentication,
                    ..
                })
            )),
            "rate_limited" => assert!(matches!(
                result,
                Ok(InvokeOutcome::Unavailable {
                    kind: UnavailableKind::RateLimited,
                    retry_after: Some(_)
                })
            )),
            "unclassified_protocol_error" => assert_eq!(
                result,
                Err(CompatibleProviderError::UnclassifiedProtocolError)
            ),
            other => panic!("unsupported fixture expectation {other}"),
        }
    }
}
